R/reshape_output.R

```diff
@@ -0,0 +1,257 @@
+    }
+  })
+  
+  # Expand the parameter estimates with additional information from the json output
+  estimates <- purrr::pmap(
+    # Combine the parameter estimates with the module name, ID, and type from 
+    # `json_list`.
+    list(
+      parameter_estimates,
+      json_list[[modules_id]][["name"]],
+      json_list[[modules_id]][["id"]],
+      json_list[[modules_id]][["type"]]
+    ),
+    ~ {
+      # Skip processing if the current estimate is NULL.
+      if (is.null(..1)) {
+        NULL
+      } else {
+        ..1 |>
+          dplyr::mutate(
+            # Add the module name from `json_list` as a new column of the current estimates.
+            module_name = ..2,
+            # Add the module ID from `json_list` as a new column.
+            module_id = ..3,
+            # Add the module type from `json_list` as a new column.
+            module_type = ..4
+          )
+      }
+    }
+  ) |>
+    # Combine all the processed tibbles into a single tibble by stacking rows.
+    dplyr::bind_rows() |>
+    # Reorder the columns to place `module_name`, `module_id`, and `module_type` at the beginning.
+    dplyr::relocate(module_name, module_id, module_type, .before = everything()) |>
+    # Remove columns `parameter_min` and `paramter_max` as we currntly don't
+    # have bounds on paramters. 
+    dplyr::select(-c(parameter_min, parameter_max))
+}
+
+#' Reshape TMB estimates
+#' 
+#' @description
+#' This function processes the TMB std and reshapes them into a structured
+#' tibble for easier analysis and manipulation.
+#'
+#' @param obj An object returned from [TMB::MakeADFun()].
+#' @param sdreport An object of the `sdreport` class as returned from
+#'   [TMB::sdreport()].
+#' @param opt An object returned from [TMB::Optimize()].
+#' @param parameter_names A character vector of parameter names. This is used to
+#'   identify the parameters in the `std` object.
+#' @return A tibble containing the reshaped estimates (i.e., parameters and
+#' derived quantities).
+reshape_tmb_estimates <- function(obj,
+                                  sdreport = NULL,
+                                  opt = NULL,
+                                  parameter_names) {
+  # Outline for the estimates table
+  estimates_outline <- dplyr::tibble(
+    # The FIMS Rcpp module
+    module_name = character(),
+    # The unique ID of the module
+    module_id = integer(),
+    # The name of the parameter or derived quantity
+    label = character(),
+    # The unique ID of the parameter
+    parameter_id = integer(),
+    # The fleet name associated with the parameter or derived quantity
+    fleet_name = character(),
+    # The age associated with the parameter or derived quantity
+    age = numeric(),
+    # The length associated with the parameter or derived quantity
+    length = numeric(),
+    # The modeled time perioed that the value pertains to
+    time = integer(),
+    # The initial value use to start the optimization procedure
```

Comment:
These four columns are NA in the output, which is fine for now but is there a plan in how to fill them in the future?

Reply:
Yes. Matthew has started adding information to the JSON output in the dev-model-families branch. We can populate these columns once the dev-model-families branch is merged. I've added a TODO comment as a reminder.